### PR TITLE
`"not source"` LocationFilter parameter

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #7789

## Feature Details
LocationFilters may now be given the parameter `"not source"`.
This parameter will cause the LocationFilter to discount the planet it is passed as the "origin" or "center", if it is given one (most cases will pass a system, but the relevant situations now pass a planet) as a possible result.
This cannot be used in conjunction with the `not` or `neighbor` modifiers.
It only has meaning for filters that are going to be used to select a planet (so, mission stopover and destination filters).

## Usage Examples
```
destination
    "not source"
    distance 0
stopover
    "not source"
    distance 0
```

## Testing Done
Created the following missions and landed on Earth a few times and checked the job board.
<summary>Test missions</summary>

<details>

```
mission " Not Source Test 1 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " Not Source Test 2 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " Not Source Test 3 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " Not Source Test 4 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " Not Source Test 5 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " Not Source Test 6 <planet>"
	job
	repeat
	source "Earth"
	destination
		distance 0
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 1 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 2 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 3 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 4 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 5 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never

mission " SHOULD NOT OFFER 6 <planet>"
	job
	repeat
	source "Earth"
	destination
		planet "Earth"
		"not source"
	to complete
		never
```
</details>

The first six jobs will offer with either Luna or Mars as the destination.
I did not see any of the other six jobs offer.

### Automated Tests Added
We don't currently test LocationFilter, and adding tests just for this feature seems excessive.

## Performance Impact
N/A

### CODE REVIEW

**Comment by warp-core** [comment]
@tibetiroka 
image

**Comment by Zitchas** [comment]
Thanks for working on this! Very much appreciated.

File: source/LocationFilter.cpp [hunk 1]
```diff
@@ -157,11 +157,14 @@ void LocationFilter::Load(const DataNode &node)
 {
 	for(const DataNode &child : node)
 	{
+		// The "not source" token may not be used with any modifiers in front of it.
+		if(child.Token(0) == "not source")
```
Comment (RisingLeaf) [inline source/LocationFilter.cpp:161]: shouldnt we have it as "not" "source"?
two separated tokens?

File: source/Mission.cpp [hunk 2]
```diff
@@ -1196,10 +1196,28 @@ Mission Mission::Instantiate(const PlayerInfo &player, const shared_ptr<Ship> &b
 		else
 			it = result.stopovers.erase(it);
 	}
+	const Planet *const sourcePlanet = player.GetPlanet();
+	auto PickPlanetStrategy = [](const Planet *sourcePlanet) -> function<const Planet *(const LocationFilter &,
```
Comment (RisingLeaf) [inline source/Mission.cpp:1200]:
```suggestion
	// Returns a function that fits the situation.
	auto PickPlanetStrategy = [](const Planet *sourcePlanet) -> function<const Planet *(const LocationFilter &,
```

When you do some super complex construct like this, how about a comment for noobs like me who have to think 5 minutes baout what this thing actually does

**Comment by warp-core** [comment]
Drafting while I work on an alternative design discussed in #8030.

**Comment by warp-core** [comment]
Closing in favour of #8055.

**Pull request closed without merging**